core/node/node_framework/src/service/mod.rs

```diff
@@ -114,55 +124,121 @@ impl ZkStackService {
+        // Add all the unconstrained tasks.
+        for task in std::mem::take(&mut self.unconstrained_tasks) {
+            let name = task.name().to_string();
+            let stop_receiver = self.stop_receiver();
+            let task_future = Box::pin(async move {
+                task.run_unconstrained(stop_receiver)
+                    .await
+                    .with_context(|| format!("Task {name} failed"))
+            });
+            tasks.push(task_future);
+        }
+        // Add all the "normal" tasks.
         for task in std::mem::take(&mut self.tasks) {
             let name = task.name().to_string();
-            let after_node_shutdown = task.after_node_shutdown();
-            let task_future = Box::pin(task.run(self.stop_receiver()));
-            let task_repr = TaskRepr {
-                name,
-                task: Some(task_future),
-                after_node_shutdown,
-            };
-            tasks.push(task_repr);
+            let stop_receiver = self.stop_receiver();
+            let task_barrier = task_barrier.clone();
+            let task_future = Box::pin(async move {
+                task.run_with_barrier(stop_receiver, task_barrier)
+                    .await
+                    .with_context(|| format!("Task {name} failed"))
+            });
+            tasks.push(task_future);
         }
         if tasks.is_empty() {
             anyhow::bail!("No tasks to run");
         }
+        // Finally, the preconditions are just futures.
+        for precondition in std::mem::take(&mut self.preconditions) {
+            let name = precondition.name().to_string();
+            let stop_receiver = self.stop_receiver();
+            let task_barrier = task_barrier.clone();
+            let task_future = Box::pin(async move {
+                precondition
+                    .check_with_barrier(stop_receiver, task_barrier)
+                    .await
+                    .with_context(|| format!("Precondition {name} failed"))
+            });
+            // Internally we use the same representation for tasks and preconditions.
+            preconditions.push(task_future);
+        }
 
         // Wiring is now complete.
         for resource in self.resources.values_mut() {
             resource.stored_resource_wired();
         }
+        tracing::info!("Wiring complete");
+
+        // Launch precondition checkers. Generally, if all of them succeed, we don't need to do anything.
+        // If any of them fails, the node have to shut down.
+        // So we run them in a detached task with an oneshot channel to report the error, if it occurs,
+        // and we'll use this channel in the same way that we use to check if any of the task exits.
+        // The difference is that preconditions, unlike tasks, are allowed to exit without causing the node
+        // to shut down.
+        let (precondition_fail_sender, precondition_fail_receiver) = oneshot::channel();
```

Comment:
Dumb question: Wouldn't retaining the handle to the task spawned below work as well?

```rust
let preconditions_result = self.runtime.spawn(async move {
    // snipped
});
let precondition_system_task = Box::pin(async move {
    match preconditions_result.await {
        // ...
    }
});
```

Reply:
Not dumb at all; it should work indeed, I'll check it.

Reply:
Actually, now that we're wrapping each precondition in its own tokio task, it seems that we don't need a detached task at all.